vm/interpreter/src/vm.rs

```diff
@@ -77,45 +105,85 @@ where
         root: Cid,
         store: DB,
         epoch: ChainEpoch,
-        rand: impl Rand + 'static,
+        rand: impl Rand + Rand_v3 + 'static,
         base_fee: TokenAmount,
         circ_supply: TokenAmount,
         reward_calc: Arc<dyn RewardCalc>,
         lb_fn: Box<dyn Fn(ChainEpoch) -> anyhow::Result<Cid>>,
-        multi_engine: &MultiEngine,
+        multi_engine_v2: &MultiEngine_v2,
+        multi_engine_v3: &MultiEngine_v3,
         chain_config: Arc<ChainConfig>,
     ) -> Result<Self, anyhow::Error> {
         let network_version = chain_config.network_version(epoch);
-        let config = NetworkConfig::new(network_version.into());
-        let engine = multi_engine.get(&config)?;
-        let mut context = config.for_epoch(epoch, root);
-        context.set_base_fee(base_fee);
-        context.set_circulating_supply(circ_supply);
-        let fvm: fvm::machine::DefaultMachine<DB, ForestExterns<DB>> =
-            fvm::machine::DefaultMachine::new(
-                &engine,
-                &context,
-                store.clone(),
-                ForestExterns::new(rand, epoch, root, lb_fn, store, chain_config),
-            )?;
-        let exec: ForestExecutor<DB> = DefaultExecutor::new(fvm);
-        Ok(VM {
-            fvm_executor: exec,
-            reward_calc,
-        })
+        if network_version >= NetworkVersion::V18 {
+            let config = NetworkConfig_v3::new(network_version.into());
+            let engine = multi_engine_v3.get(&config)?;
+            // The UNIX timestamp (in seconds) of the current tipset.
+            // XXX: How do we get this?
```

Comment:
I believe like so https://github.com/filecoin-project/lotus/blob/release/v1.20.0/chain/types/tipset.go#L214-L219

So grabbing the smallest timestamp from the headers of the current tipset's blocks

Reply:
Oof, not gonna be fun to write that up. Mind taking a stab at it?

Reply:
It turned out not fun indeed; we already had this method implemented.